src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+            if (instanceGenericConstraint != null)
+            {
+                var result = DuckType.GetOrCreateProxyType(instanceGenericConstraint, targetType);
+                instanceProxyType = result.ProxyType;
+                callGenericTypes.Add(instanceProxyType);
+            }
+            else
+            {
+                callGenericTypes.Add(targetType);
+            }
+
+            DynamicMethod callMethod = new DynamicMethod(
+                     $"{onMethodBeginMethodInfo.DeclaringType.Name}.{onMethodBeginMethodInfo.Name}",
+                     typeof(CallTargetState),
+                     new Type[] { targetType, typeof(object[]) },
+                     onMethodBeginMethodInfo.Module,
+                     true);
+
+            ILGenerator ilWriter = callMethod.GetILGenerator();
+
+            // Load the instance if is needed
+            if (mustLoadInstance)
+            {
+                ilWriter.Emit(OpCodes.Ldarg_0);
+
+                if (instanceGenericConstraint != null)
+                {
+                    ConstructorInfo ctor = instanceProxyType.GetConstructors()[0];
+                    if (targetType.IsValueType && !ctor.GetParameters()[0].ParameterType.IsValueType)
+                    {
+                        ilWriter.Emit(OpCodes.Box, targetType);
+                    }
+
+                    ilWriter.Emit(OpCodes.Newobj, ctor);
+                }
+            }
+
+            // Load arguments
+            for (var i = mustLoadInstance ? 1 : 0; i < onMethodBeginParameters.Length; i++)
+            {
+                Type targetParameterType = onMethodBeginParameters[i].ParameterType;
+                Type targetParameterTypeConstraint = null;
+
+                if (targetParameterType.IsGenericParameter)
+                {
+                    targetParameterType = genericArgumentsTypes[targetParameterType.GenericParameterPosition];
+
+                    targetParameterTypeConstraint = targetParameterType.GetGenericParameterConstraints().FirstOrDefault(pType => pType != typeof(IDuckType));
+                    if (targetParameterTypeConstraint is null)
+                    {
+                        callGenericTypes.Add(typeof(object));
+                    }
+                    else
+                    {
+                        targetParameterType = targetParameterTypeConstraint;
+                        callGenericTypes.Add(targetParameterTypeConstraint);
+                    }
+                }
+
+                ilWriter.Emit(OpCodes.Ldarg_1);
+                WriteIntValue(ilWriter, i - (mustLoadInstance ? 1 : 0));
+                ilWriter.Emit(OpCodes.Ldelem_Ref);
+
+                if (targetParameterTypeConstraint != null)
+                {
+                    ilWriter.EmitCall(OpCodes.Call, ConvertTypeMethodInfo.MakeGenericMethod(targetParameterTypeConstraint), null);
+                }
+
+                if (targetParameterType.IsValueType)
```

Comment:
Can this be an `else if` instead of an `if`, without affecting the program behavior?

Reply:
yes, indeed it should be an `else if` ConvertType can return a struct proxy of a struct instance, and that `Unbox_Any` could lead to an invalid boxing there.

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/04ae61a340872f36f5e93929e69b26305bc991dd